File: phix/src/main/kotlin/subst/theoria/flow/RMFlow.kt
package subst.theoria.flow

import kotlinx.coroutines.flow.Flow
import subst.theoria.judgment.*
import subst.theoria.phi.executor.FlowExecutor
import subst.theoria.phi.executor.JudgmentEvent
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

// Host: restriction-modification system을 가진 숙주.
data class Host(
    val recognitionStrength: Double,
    val methylationLevel: Double,
    var damage: Double = 0.0
)

data class Phage(
    var resistance: Double,
    var replicationRate: Double
)

data class InteractionState(
    var phageLoad: Double = 1.0,
    var time: Int = 0
)

// Restriction-Modification system과 phage 간의 확률적 동역학 경쟁을 PhiJudgment으로 변환
class RMFlow(
    private val host: Host,
    private val phage: Phage,
    private val state: InteractionState,
    private val maxSteps: Int = 50
) {

    /**
     * - 결과를 PhiJudgment으로 구성한다.
     * - 실제 시간 반복은 while 루프로 계산되지만,
     * - 엔진에는 오직 위상 이벤트만 전달된다.
     */
    private fun buildPhiJudgment(): PhiJudgment {
        val decls = mutableListOf<PhiDecl>()

        /**
         * 반복 조건:
         * - 시간 제한
         * - phage 완전 제거 전
         * - host 완전 붕괴 전
         */
        while (
            state.time < maxSteps &&
            state.phageLoad > 0.01 &&
            host.damage < 1.0
        ) {
            state.time++

            /**
             * Recognition pressure:
             * - 숙주의 인식 능력 × phage 회피 실패 확률
             * > phage resistance가 높을수록 recognitionPressure는 감소.
             */
            val recognitionPressure = host.recognitionStrength * (1.0 - phage.resistance)

            /**
             * Effective cleavage:
             * - recognition에서 methylation 보호 효과를 뺀 값.
             * - 실제 절단 가능성의 근사값.
             */
            val effectiveCleavage = recognitionPressure - host.methylationLevel
            val cleavageProbability = max(0.0, min(1.0, effectiveCleavage))
            val cleavage = Random.nextDouble() < cleavageProbability
            if (cleavage) {
                /**
                 * Restriction event:
                 * - phage genome 절단 → 개체수 감소
                 * - 이는 위상적으로 "host-defense 우세" 상태.
                 */
                state.phageLoad *= 0.5

                decls += FlowDecl(
                    from = Psi("cleavage"),
                    to = Phi("phage-reduced"),
                    bound = Bound("host-defense")
                )

            } else {

                /**
                 * Replication event:
                 * - 절단 실패 → phage 증식
                 * - 이는 위상적으로 "viral-dynamics 우세" 상태.
                 */
                state.phageLoad *= phage.replicationRate

                decls += FlowDecl(
                    from = Psi("replication"),
                    to = Phi("phage-expanded"),
                    bound = Bound("viral-dynamics")
                )
            }

            /**
             * 숙주 손상 누적:
             * - phageLoad가 클수록 세포 손상 증가.
             * - 이는 장기적으로 lytic collapse를 유도.
             */
            host.damage += state.phageLoad * 0.01
        }

        /**
         * 시스템의 종결 상태 결정:
         * - Restriction victory
         * - Lytic collapse
         * - Stable coexistence
         */
        val outcome =
            when {
                state.phageLoad < 0.01 -> "restriction-victory"
                host.damage > 1.0 -> "lytic-collapse"
                else -> "stable-coexistence"
            }

        // CollapseDecl: 연속적 경쟁이 하나의 거시적 결과 상태로 수렴하는 지점.
        decls += CollapseDecl(
            sources = listOf(Psi("rm-dynamics")),
            bound = Phi(outcome)
        )

        return PhiJudgment(
            steps = listOf(
                StepDecl(
                    name = "rm.interaction",
                    body = decls
                )
            )
        )
    }

    // 외부에서는 단순히 Flow<PhaseEvent>로 소비 가능.
    fun execute(): Flow<JudgmentEvent> {
        val judgment = buildPhiJudgment()
        return FlowExecutor().execute(judgment)
    }
}

File: xphi/flow/rmflow.py
# xphi.flow.rmflow
## @lineage: meta.xphi.rmflow
## @lineage: xphi.rmflow
## @lineage: meta.ops.xphi.rmflow
import argparse
import asyncio
import random
from dataclasses import dataclass
from typing import List, AsyncGenerator, AsyncIterable
from phase.plane.emitter import get_emitter
from arch.contract.registry.unified import contract

log = get_emitter("xphi.rmflow")

@dataclass
class Psi:
    name: str

@dataclass
class Phi:
    name: str

@dataclass
class Bound:
    name: str

class PhiDecl:
    pass

@dataclass
class FlowDecl(PhiDecl):
    from_: Psi
    to: Phi
    bound: Bound

@dataclass
class CollapseDecl(PhiDecl):
    sources: List[Psi]
    bound: Phi

@dataclass
class StreamStepDecl:
    name: str
    body_stream: AsyncIterable[PhiDecl]

@dataclass
class StreamJudgment:
    steps: List[StreamStepDecl]

@dataclass
class Host:
    recognition_strength: float
    methylation_level: float
    damage: float = 0.0

@dataclass
class Phase:
    resistance: float
    replication_rate: float

@dataclass
class InteractionState:
    phase_load: float = 1.0
    time: int = 0

class JudgmentEvent:
    pass

@dataclass
class FlowEvent(JudgmentEvent):
    psi: Psi
    phi: Phi
    boundary: Bound
    phase: str

@dataclass
class CollapseEvent(JudgmentEvent):
    sources: List[Psi]
    surface: Phi
    phase: str

class ExecutionContext:
    def __init__(self):
        self.current_phase: str = "INIT"

    def phase(self, name: str):
        self.current_phase = name

    async def run_stream_step(self, step: StreamStepDecl) -> AsyncGenerator[JudgmentEvent, None]:
        self.phase(step.name)
        async for decl in step.body_stream:
            if isinstance(decl, FlowDecl):
                yield self.emit_flow(psi=decl.from_, phi=decl.to, bound=decl.bound)
            elif isinstance(decl, CollapseDecl):
                yield self.emit_collapse(sources=decl.sources, bound=decl.bound)
            else:
                log.warning(f"[runStep] unknown decl: {decl}")

    def emit_flow(self, psi: Psi, phi: Phi, bound: Bound) -> FlowEvent:
        return FlowEvent(psi=psi, phi=phi, boundary=bound, phase=self.current_phase)

    def emit_collapse(self, sources: List[Psi], bound: Phi) -> CollapseEvent:
        return CollapseEvent(sources=sources, surface=bound, phase=self.current_phase)

class FlowExecutor:
    async def execute(self, judgment: StreamJudgment) -> AsyncGenerator[JudgmentEvent, None]:
        ctx = ExecutionContext()
        for step in judgment.steps:
            async for event in ctx.run_stream_step(step):
                yield event

class RMFlow:
    def __init__(self, host: Host, phase: Phase, state: InteractionState, max_steps: int = 50):
        self.host = host
        self.phase = phase
        self.state = state
        self.max_steps = max_steps

    async def _generate_interaction(self) -> AsyncGenerator[PhiDecl, None]:
        """시뮬레이션 로직을 실시간으로 계산하며 위상(Decl)을 방출합니다."""
        while (
            self.state.time < self.max_steps and
            self.state.phase_load > 0.01 and
            self.host.damage < 1.0
        ):
            self.state.time += 1

            # Recognition pressure & Effective cleavage 계산
            recognition_pressure = self.host.recognition_strength * (1.0 - self.phase.resistance)
            effective_cleavage = recognition_pressure - self.host.methylation_level
            cleavage_probability = max(0.0, min(1.0, effective_cleavage))
            
            if random.random() < cleavage_probability:
                self.state.phase_load *= 0.5
                yield FlowDecl(from_=Psi("cleavage"), to=Phi("phase-reduced"), bound=Bound("host-defense"))
            else:
                self.state.phase_load *= self.phase.replication_rate
                yield FlowDecl(from_=Psi("replication"), to=Phi("phase-expanded"), bound=Bound("viral-dynamics"))

            self.host.damage += self.state.phase_load * 0.01
            await asyncio.sleep(0)

        if self.state.phase_load < 0.01:
            outcome = "restriction-victory"
        elif self.host.damage > 1.0:
            outcome = "lytic-collapse"
        else:
            outcome = "stable-coexistence"

        yield CollapseDecl(sources=[Psi("rm-dynamics")], bound=Phi(outcome))

    def _build_stream_judgment(self) -> StreamJudgment:
        """선언적 구조(Judgment)를 반환하지만, 내부는 실시간 스트림(generator)을 참조합니다."""
        return StreamJudgment(
            steps=[
                StreamStepDecl(
                    name="rm.interaction",
                    body_stream=self._generate_interaction()
                )
            ]
        )

    async def execute(self) -> AsyncGenerator[JudgmentEvent, None]:
        judgment = self._build_stream_judgment()
        async for event in FlowExecutor().execute(judgment):
            yield event

@contract.flow(name="xphi.rmflow", entry="rmflow_entry")
def rmflow_entry(cli_args: list = None, **payload):
    """외부 자극(Psi)의 초기 에너지(payload)를 받아 RMFlow 위상을 인스턴스화"""
    ## Payload가 있으면 사용하고, 없으면 기본값 사용
    rec_str = payload.get("recognition_strength", 0.8)
    rep_rate = payload.get("replication_rate", 1.5)
    steps = payload.get("steps", 50)

    host = Host(recognition_strength=rec_str, methylation_level=0.2)
    phase = Phase(resistance=0.3, replication_rate=rep_rate)
    state = InteractionState()
    return RMFlow(host, phase, state, max_steps=steps)

if __name__ == "__main__":
    from cognitive.flow.executor import dispatch_flow_cli
    dispatch_flow_cli(
        command_name="rmflow", 
        entry_func=rmflow_entry, 
        file_path=__file__
    )